### PR TITLE
Support role as granted authority

### PR DESCRIPTION
**Pull request by yiliuTo**
For feature request in #19250 , this pr is aimed to support generating granted authorities from "roles" claim in id-token for web application.

### CODE REVIEW

File: sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-webapp/src/main/java/com/azure/spring/sample/aad/controller/RoleController.java [hunk 1]
```diff
@@ -23,4 +23,11 @@ public String group1() {
     public String group2() {
         return "group2 message";
     }
+
+    @GetMapping("Admin")
+    @ResponseBody
+    @PreAuthorize("hasRole('ROLE_Admin')")
```
Comment (chenrujun) [inline sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-webapp/src/main/java/com/azure/spring/sample/aad/controller/RoleController.java:29]: According to email `FastTrack for Azure - Feedback on v3 AAD Azure Spring starter`, it should be `APPROLE_Admin`

File: sdk/spring/azure-spring-boot-starter-active-directory/README.md [hunk 2]
```diff
@@ -368,6 +368,40 @@ After the scopes have been consented, AAD server will remember that this user ha
 the permission to the web application. So incremental consent will not happen anymore after user 
 consented.
 
+### Enhanced Azure Active Directory features
```
Comment (chenrujun) [inline sdk/spring/azure-spring-boot-starter-active-directory/README.md:371]: `Enhanced Azure Active Directory features` -> `Advanced  features`

File: sdk/spring/azure-spring-boot-starter-active-directory/README.md [hunk 3]
```diff
@@ -368,6 +368,40 @@ After the scopes have been consented, AAD server will remember that this user ha
...
+          ],
+          "displayName": "Admin",
+          "id": "2fa848d0-8054-4e11-8c73-7af5f1171001",
+          "isEnabled": true,
+          "description": "Full admin access",
+          "value": "Admin"
+         }
+      ]
+    ```
+* Step 2: Write Java code:
```
Comment (chenrujun) [inline sdk/spring/azure-spring-boot-starter-active-directory/README.md:392]: Add tag to the code.
Example:
```
<!-- embedme ../azure-spring-boot/src/samples/java/com/azure/spring/aad/AADOAuth2LoginConfigSample.java#L18-L29 -->
```

Same to all other places in this file.

File: sdk/spring/azure-spring-boot-starter-active-directory/README.md [hunk 4]
```diff
@@ -368,6 +368,40 @@ After the scopes have been consented, AAD server will remember that this user ha
...
 consented.
 
+### Enhanced Azure Active Directory features
+#### Support access control by id token in web application 
+This starter supports creating `GrantedAuthority` from id_token's `roles` claim to allow using `id_token` for authorization in web application. Developers can use the
+`appRoles` feature of Azure Active Directory to create `roles` claim and implement access control. 
+Follow the guide to 
+[add app roles in your application and assign to users or groups](https://docs.microsoft.com/azure/active-directory/develop/howto-add-app-roles-in-azure-ad-apps).
+* Step 1: Add below `appRoles` in your application's manifest:
+    ```
```
Comment (chenrujun) [inline sdk/spring/azure-spring-boot-starter-active-directory/README.md:378]: nit: delete preceding white space.

File: sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-webapp/README.md [hunk 5]
```diff
@@ -22,6 +22,28 @@ See [Register app], [Grant scoped permission] for more information about web app
 ### Configure groups for sign in user
 In order to try the authorization action with this sample with minimum effort, [configure the user and groups in Azure Active Directory], configure the user with `group1`.
 
+### Configure AppRole for web application
+If you want to use `id_token` for authorization, we can use 
+`appRoles` feature which AAD provides and create `GrantedAuthority` from id_token's `roles` claim to implement access control. 
+Follow the guide to 
+[add app roles in your application](https://docs.microsoft.com/azure/active-directory/develop/howto-add-app-roles-in-azure-ad-apps).
+1. In this example you need to create following `appRoles` in your application's manifest:
+    ```
```
Comment (chenrujun) [inline sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-webapp/README.md:31]: nit: delete preceding white space.

File: sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/aad/webapp/AADOAuth2UserService.java [hunk 6]
```diff
@@ -62,6 +67,13 @@ public OidcUser loadUser(OidcUserRequest userRequest) throws OAuth2Authenticatio
             return (DefaultOidcUser) session.getAttribute(DEFAULT_OIDC_USER);
         }
 
+        if (idToken.containsClaim(ROLES)) {
+            roles = idToken.getClaimAsStringList(ROLES)
+                           .stream()
+                           .filter(s -> StringUtils.hasText(s))
+                           .map(role -> ROLE_PREFIX + role)
```
Comment (chenrujun) [inline sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/aad/webapp/AADOAuth2UserService.java:74]: `APPROLE_PREFIX`

**Review by chenrujun — CHANGES_REQUESTED** [review]

File: sdk/spring/azure-spring-boot/src/samples/java/com/azure/spring/aad/CustomWebServerSecurityConfig.java [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+import org.springframework.security.config.annotation.web.builders.HttpSecurity;
+import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
+
+@EnableWebSecurity
+@EnableGlobalMethodSecurity(prePostEnabled = true)
+public class CustomWebServerSecurityConfig extends AADResourceServerWebSecurityConfigurerAdapter {
+    @Override
+    protected void configure(HttpSecurity http) throws Exception {
+        super.configure(http);
+        // Do some custom configuration
```
Comment (chenrujun) [inline sdk/spring/azure-spring-boot/src/samples/java/com/azure/spring/aad/CustomWebServerSecurityConfig.java:14]: Can we use the code in sample project? `azure-spring-boot-sample-active-directory-webapp`.

File: sdk/spring/azure-spring-boot/src/samples/java/com/azure/spring/aad/JsonMapper.java [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+package com.azure.spring.aad;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.SerializationFeature;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.security.oauth2.client.OAuth2AuthorizedClient;
+
+public class JsonMapper {
```
Comment (chenrujun) [inline sdk/spring/azure-spring-boot/src/samples/java/com/azure/spring/aad/JsonMapper.java:13]: Same here.

File: sdk/spring/azure-spring-boot/src/samples/java/com/azure/spring/aad/SampleController.java [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+import org.springframework.security.oauth2.client.OAuth2AuthorizedClient;
+import org.springframework.security.oauth2.client.annotation.RegisteredOAuth2AuthorizedClient;
+import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.ResponseBody;
+
+import static com.azure.spring.aad.JsonMapper.toJsonString;
+
+@Controller
+public class SampleController {
```
Comment (chenrujun) [inline sdk/spring/azure-spring-boot/src/samples/java/com/azure/spring/aad/SampleController.java:16]: Here.

**Review by chenrujun — CHANGES_REQUESTED** [review]

File: sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/aad/webapp/AADOAuth2UserService.java [hunk 10]
```diff
@@ -62,6 +68,13 @@ public OidcUser loadUser(OidcUserRequest userRequest) throws OAuth2Authenticatio
             return (DefaultOidcUser) session.getAttribute(DEFAULT_OIDC_USER);
         }
 
+        if (idToken.containsClaim(ROLES)) {
+            roles = idToken.getClaimAsStringList(ROLES)
```
Comment (totakura) [inline sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/aad/webapp/AADOAuth2UserService.java:72]: Please also check for NULL. According to [docs](https://docs.spring.io/spring-security/site/docs/current/api/org/springframework/security/oauth2/core/ClaimAccessor.html#getClaimAsStringList-java.lang.String-) this method could return NULL if the claim doesn't exist or is not a list.

Comment (chenrujun) [reply sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/aad/webapp/AADOAuth2UserService.java:72]: @yiliuTo How about write like this:
```
        Optional.ofNullable(idToken)
                .map(token -> token.getClaimAsStringList(ROLES))
                .map(Collection::stream)
                .orElseGet(Stream::empty)
                .filter(StringUtils::hasText)
                .map(role -> APPROLE_PREFIX + role)
                .forEach(roles::add);
```

And please add test about:
 1. claim does not exist.
 2. cannot be assigned to a list

File: sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-webapp/README.md [hunk 11]
```diff
@@ -22,6 +22,30 @@ See [Register app], [Grant scoped permission] for more information about web app
 ### Configure groups for sign in user
 In order to try the authorization action with this sample with minimum effort, [configure the user and groups in Azure Active Directory], configure the user with `group1`.
 
+### Configure AppRole for web application
+If you want to use `id_token` for authorization, we can use 
+`appRoles` feature of AAD to generate id_token's `roles` claim and then create `GrantedAuthority` from `roles` to implement access control. 
+Note the `roles` claim generated from `appRoles` is decorated with prefix `APPROLE`.
```
Comment (chenrujun) [inline sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-webapp/README.md:28]: 1. `APPROLE` -> `APPROLE_`
2. Adjust the line breaks:
image

File: sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-webapp/src/main/java/com/azure/spring/sample/aad/controller/RoleController.java [hunk 12]
```diff
@@ -23,4 +23,11 @@ public String group1() {
     public String group2() {
         return "group2 message";
     }
+
+    @GetMapping("Admin")
+    @ResponseBody
+    @PreAuthorize("hasAuthority('APPROLE_Admin')")
```
Comment (chenrujun) [inline sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-webapp/src/main/java/com/azure/spring/sample/aad/controller/RoleController.java:29]: Maybe it should move it to a new Controller named `AuthorityController`?

File: sdk/spring/azure-spring-boot-starter-active-directory/CHANGELOG.md [hunk 13]
```diff
@@ -1,7 +1,8 @@
 # Release History
 
 ## 3.3.0-beta.1 (Unreleased)
-
+### New Features
```
Comment (chenrujun) [inline sdk/spring/azure-spring-boot-starter-active-directory/CHANGELOG.md:4]: Need to update after 3.3.0 released.

File: sdk/spring/azure-spring-boot-starter-active-directory/README.md [hunk 14]
```diff
@@ -40,57 +40,57 @@ This scenario uses the [The OAuth 2.0 authorization code grant] flow to login in
...
-        http.authorizeRequests()
-            .antMatchers("/login").permitAll()
-            .anyRequest().authenticated();
+    `AADWebSecurityConfigurerAdapter` contains necessary web security configuration for **aad-starter**.
+
+     (A). `DefaultAADWebSecurityConfigurerAdapter` is configured automatically if you not provide one.
+    
+     (B). You can provide one by extending `AADWebSecurityConfigurerAdapter` and call `super.configure(http)` explicitly 
+    in the `configure(HttpSecurity http)` function. Here is an example:
+    <!-- embedme ../azure-spring-boot/src/samples/java/com/azure/spring/aad/AADOAuth2LoginConfigSample.java#L18-L29 -->
```
Comment (chenrujun) [inline sdk/spring/azure-spring-boot-starter-active-directory/README.md:79]: Could you please update it to related code in sample project, and delete in `src/sample/java`.

File: sdk/spring/azure-spring-boot-starter-active-directory/README.md [hunk 15]
```diff
@@ -274,100 +286,129 @@ Here are some examples about how to use these properties:
...
-@EnableGlobalMethodSecurity(prePostEnabled = true)
-public class AADOAuth2LoginConfigSample extends AADWebSecurityConfigurerAdapter {
-
-    @Override
-    protected void configure(HttpSecurity http) throws Exception {
-        super.configure(http);
-        http.authorizeRequests()
-            .antMatchers("/login").permitAll()
-            .anyRequest().authenticated();
+    <!-- embedme ../azure-spring-boot/src/samples/java/com/azure/spring/aad/AADOAuth2LoginConfigSample.java#L18-L29 -->
```
Comment (chenrujun) [inline sdk/spring/azure-spring-boot-starter-active-directory/README.md:308]: Same here.

File: sdk/spring/azure-spring-boot-starter-active-directory/README.md [hunk 16]
```diff
@@ -274,100 +286,129 @@ Here are some examples about how to use these properties:
...
+    After the scopes have been consented, AAD server will remember that this user has already granted 
+    the permission to the web application. So incremental consent will not happen anymore after user 
+    consented.
+
+### Advanced features
+
+#### Support access control by id token in web application 
+
+This starter supports creating `GrantedAuthority` from id_token's `roles` claim to allow using `id_token` for authorization in web application. Developers can use the
+`appRoles` feature of Azure Active Directory to create `roles` claim and implement access control. Note the `roles` claim generated from `appRoles` is decorated with prefix `APPROLE`.
```
Comment (chenrujun) [inline sdk/spring/azure-spring-boot-starter-active-directory/README.md:383]: `APPROLE` -> `APPROLE_`

File: sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/aad/webapp/AADOAuth2UserService.java [hunk 17]
```diff
@@ -54,6 +62,8 @@ public AADOAuth2UserService(
     public OidcUser loadUser(OidcUserRequest userRequest) throws OAuth2AuthenticationException {
         // Delegate to the default implementation for loading a user
         OidcUser oidcUser = oidcUserService.loadUser(userRequest);
+        OidcIdToken idToken = oidcUser.getIdToken();
+        Set<String> roles = new HashSet<>();
```
Comment (chenrujun) [inline sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/aad/webapp/AADOAuth2UserService.java:66]: `roles` -> `authorityStrings`

File: sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/aad/webapp/AADOAuth2UserService.java [hunk 18]
```diff
@@ -62,19 +72,12 @@ public OidcUser loadUser(OidcUserRequest userRequest) throws OAuth2Authenticatio
...
-                                     .map(graphClient::getGroupsFromGraph)
-                                     .orElseGet(Collections::emptySet);
-        Set<String> groupRoles = groups.stream()
-                                       .filter(properties::isAllowedGroup)
-                                       .map(group -> ROLE_PREFIX + group)
-                                       .collect(Collectors.toSet());
-        Set<SimpleGrantedAuthority> authorities = groupRoles.stream()
-                                                            .map(SimpleGrantedAuthority::new)
-                                                            .collect(Collectors.toSet());
+        extractRolesFromIdToken(idToken, roles);
```
Comment (chenrujun) [inline sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/aad/webapp/AADOAuth2UserService.java:75]:
```
authorityStrings.addAll(getAuthorityFromIdtoken(idToken)).
```

File: sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/aad/webapp/AADOAuth2UserService.java [hunk 19]
```diff
@@ -62,19 +72,12 @@ public OidcUser loadUser(OidcUserRequest userRequest) throws OAuth2Authenticatio
...
+        extractGroupRolesFromAccessToken(userRequest.getAccessToken(), roles);
```
Comment (chenrujun) [inline sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/aad/webapp/AADOAuth2UserService.java:76]: Same here.

**Review by chenrujun — CHANGES_REQUESTED** [review]

**Review by chenrujun — APPROVED** [review]
LGTM.

**Pull request merged**